### PR TITLE
[BB-3670] Include only new instances in send_trial_instances_report task

### PR DESCRIPTION
**Pull request by spokerman12**
**Description**

We're fulfilling the following user story:

> "As the person converting Ocim trial users, I need an autogenerated monthly report with a list of trial users for each month. "

**Supporting information**

- [Jira ticket](https://tasks.opencraft.com/browse/BB-3670)

**Testing instructions**

1. Set up a local OCIM environment using this branch
2. Create instances and appservers
3. Edit their created dates with any method available
4. Run `send_trial_instances_report` on `reports/tasks.py` or
5. Run `./manage.py test reports.tests.ReportTaskTestCase` inside vagrant VM

**Snapshots:**

None

**Further considerations:**

- Initially we enabled a flag for the command, but we updated the feature to reflect current requirements as discussed in the review comments.

Reviewer:
[ ] @lgp171188

### CODE REVIEW

File: instance/management/commands/instance_statistics_csv.py [hunk 1]
```diff
@@ -98,6 +98,13 @@ def add_arguments(self, parser):
             default=None,
             help='Path to the output file of the new CSV. Leave blank to use stdout.'
         )
+        parser.add_argument(
```
Comment (lgp171188) [inline instance/management/commands/instance_statistics_csv.py:101]: @spokerman12, the management command should be left as-is and the code [here](https://github.com/open-craft/opencraft/blob/master/reports/tasks.py) should be updated to do the necessary filtering.

Comment (spokerman12) [reply instance/management/commands/instance_statistics_csv.py:101]: I don't understand, on this [comment](https://tasks.opencraft.com/browse/BB-3670?focusedCommentId=188674&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-188674) you said 

> [implement] a flag to the existing command that takes in a date range and filters the instances down to that

- Are domains always required, then?
- Should I remove the --created flag?
- Should I create another db_periodic_task or edit the current send_trial_instances_report? The name is confusing as it suggests I should edit it.

Comment (lgp171188) [reply instance/management/commands/instance_statistics_csv.py:101]: @spokerman12, yes, I said what you have quoted above. But I feel that it is better to leave the command as-is and generic enough and update the periodic task instead to filter and pass only the instances created in the most recent completed month. Does this make sense or do you have any questions?

Comment (spokerman12) [reply instance/management/commands/instance_statistics_csv.py]: Makes sense! I implemented the changes suggested plus a test.

File: reports/tasks.py [hunk 2]
```diff
@@ -46,27 +46,19 @@
         day_of_week=TRIAL_INSTANCES_REPORT_SCHEDULE_DAY_OF_WEEK
     )
 )
-def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIPIENTS):
+def send_trial_instances_report(
+        recipients=settings.TRIAL_INSTANCES_REPORT_RECIPIENTS):
     """
-    Generate and send a trial instance data report
+    Generate and send a new instance data report for the past month
```
Comment (lgp171188) [inline reports/tasks.py:52]: @spokerman12, nit: the new instances are still trial instances :)

File: reports/tasks.py [hunk 3]
```diff
@@ -46,27 +46,19 @@
         day_of_week=TRIAL_INSTANCES_REPORT_SCHEDULE_DAY_OF_WEEK
     )
 )
-def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIPIENTS):
+def send_trial_instances_report(
+        recipients=settings.TRIAL_INSTANCES_REPORT_RECIPIENTS):
```
Comment (lgp171188) [inline reports/tasks.py:50]: @spokerman12, is there a reason you changed the formatting of this line? The old one met the code quality checks and didn't cause any issues before.

And if there is some reason to change it, you should format it according to our coding standards, like what you had to do for the django-password-validators PR.

File: reports/tasks.py [hunk 4]
```diff
@@ -46,27 +46,19 @@
...
-    Generate and send a trial instance data report
+    Generate and send a new instance data report for the past month
 
     This task runs on the first of every month at 2AM
     """
 
     if not recipients:
-        logger.warning('No recipients listed for Trial Instances Report. It will not be generated.')
+        logger.warning(
+            'No recipients listed for Trial Instances Report. It will not be generated.')
```
Comment (lgp171188) [inline reports/tasks.py:59]: @spokerman12, similar concerns as https://github.com/open-craft/opencraft/pull/711/files#r579396180

File: reports/tasks.py [hunk 5]
```diff
@@ -76,6 +68,21 @@ def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIP
...
+    # Get instances with an active betatestapplication,
+    # whose InstanceReference has an active OpenEdXAppServer
+    # and was created last month
+    instances = OpenEdXInstance.objects.exclude(
+        betatestapplication__isnull=True
+    ).filter(
+        ref_set__openedxappserver_set___is_active=True
+    ).filter(
+        ref_set__created__range=[
+            beginning_of_last_month.strftime("%Y-%m-%d"),
```
Comment (lgp171188) [inline reports/tasks.py:80]: @spokerman12, since this is just a time range, how much effort would it require to make this configurable instead of hard-coding it to the last completed month?

File: reports/tasks.py [hunk 6]
```diff
@@ -76,6 +68,21 @@ def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIP
...
+            end_of_last_month.strftime("%Y-%m-%d")
+        ])
```
Comment (lgp171188) [inline reports/tasks.py:82]: @spokerman12, nit: closing parenthesis formatting.

File: reports/tasks.py [hunk 7]
```diff
@@ -92,7 +99,8 @@ def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIP
         )
     except SystemExit:
         # The command exited prematurely. Send failure email to recipients
-        logger.error('`instance_statistics_csv` command failed. Sending notification email')
+        logger.error(
+            '`instance_statistics_csv` command failed. Sending notification email')
```
Comment (lgp171188) [inline reports/tasks.py:103]: @spokerman12, similar concerns as https://github.com/open-craft/opencraft/pull/711/files#r579396180

File: reports/tasks.py [hunk 8]
```diff
@@ -76,6 +68,21 @@ def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIP
...
+        betatestapplication__isnull=True
+    ).filter(
+        ref_set__openedxappserver_set___is_active=True
+    ).filter(
+        ref_set__created__range=[
+            beginning_of_last_month.strftime("%Y-%m-%d"),
+            end_of_last_month.strftime("%Y-%m-%d")
+        ])
+    domains = [
+        instance.external_lms_domain or instance.internal_lms_domain for instance in instances
```
Comment (lgp171188) [inline reports/tasks.py:84]: @spokerman12, I think `instance.domain` will always contain the right value and so this `or` condition may not be needed.

File: reports/tests.py [hunk 9]
```diff
@@ -528,3 +533,110 @@ def _generate_invoice_date(year=datetime.now().year, month=1, this_month=False):
...
+
+        user1, _ = get_user_model().objects.get_or_create(
+            username='test1', email='test1@example.com')
+        user2, _ = get_user_model().objects.get_or_create(
+            username='test2', email='test2@example.com')
+        user3, _ = get_user_model().objects.get_or_create(
+            username='test3', email='test3@example.com')
+
+        with freeze_time(date1):
+            instance1 = OpenEdXInstanceFactory(
```
Comment (lgp171188) [inline reports/tests.py:581]: @spokerman12, the repeated code below for creating multiple users can be refactored into a separate function or a factory function, no?

File: reports/tests.py [hunk 10]
```diff
@@ -528,3 +533,110 @@ def _generate_invoice_date(year=datetime.now().year, month=1, this_month=False):
...
+                successfully_provisioned=True, betatestapplication=BetaTestApplication())
+            make_test_appserver(instance=instance1, is_active=True)
+            BetaTestApplication.objects.create(
+                user=user1,
+                public_contact_email='publicemail1@example.com',
+                subdomain='betatestdomain1',
+                instance_name=instance1.name,
+                status=BetaTestApplication.PENDING,
```
Comment (lgp171188) [inline reports/tests.py:589]: @spokerman12, I don't think `PENDING` is the correct state. Once an appserver is created, I think the status changes to "approved".

File: reports/tasks.py (hunk 2 above)
Comment (spokerman12) [reply reports/tasks.py:52]: Understood. Changed like so.

File: reports/tasks.py (hunk 3 above)
Comment (spokerman12) [reply reports/tasks.py:50]: I ran this through autopep8, but it's true that while we should adhere to PEP8 per the handbook, the 79 character per line limit is changed to 120.

File: reports/tasks.py (hunk 4 above)
Comment (spokerman12) [reply reports/tasks.py:59]: I ran this through autopep8, but it's true that while we should adhere to PEP8 per the handbook, the 79 character per line limit is changed to 120.

File: reports/tasks.py (hunk 5 above)
Comment (spokerman12) [reply reports/tasks.py:80]: It'd be easy, just adding more arguments to send_trial_instances_report.
But right now I don't think it's convenient to increase the scope as we still have tasks left; I'd be happy to do it on the next sprint, though.

File: reports/tasks.py (hunk 6 above)
Comment (spokerman12) [reply reports/tasks.py:82]: Changed according to previous suggestions.

File: reports/tasks.py (hunk 8 above)
Comment (spokerman12) [reply reports/tasks.py:84]: This part of the logic was already implemented, so I wouldn't want to rework it unnecessarily; plus, I don't think the command is tested thoroughly. As suggested before, I'd be more than happy to work on this on a different task.

File: reports/tasks.py (hunk 7 above)
Comment (spokerman12) [reply reports/tasks.py:103]: Corrected as suggested.

File: reports/tests.py (hunk 9 above)
Comment (spokerman12) [reply reports/tests.py:581]: Because it's implied that there's an ordering  user1 < user2 < user3, instance1 < instance2 < instance3 ... etc, because we're adding a singular feature I prefer the test to be explicit (or "don't make me think," as suggested in the handbook).

I'd add the factory function if we were adding more tests that could make use of it, but in this case I don't think it's necessary.

File: reports/tests.py (hunk 10 above)
Comment (spokerman12) [reply reports/tests.py:589]: The command only wants that the referred BetaTestApplication is not null, but I'll change it to Approved to reflect real scenarios.

File: reports/tasks.py [hunk 11]
```diff
@@ -76,6 +66,22 @@ def send_trial_instances_report(recipients=settings.TRIAL_INSTANCES_REPORT_RECIP
...
+    # whose InstanceReference has an active OpenEdXAppServer
+    # and was created last month
+    instances = OpenEdXInstance.objects.exclude(
+        betatestapplication__isnull=True
+    ).filter(
+        ref_set__openedxappserver_set___is_active=True
+    ).filter(
+        ref_set__created__range=[
+            beginning_of_last_month.strftime("%Y-%m-%d"),
+            end_of_last_month.strftime("%Y-%m-%d")
```
Comment (lgp171188) [inline reports/tasks.py:79]: @spokerman12, this has the off-by-one error as mentioned [here](https://docs.djangoproject.com/en/3.1/ref/models/querysets/#range). Can you fix it and also write a test case for this scenario?

File: reports/tests.py [hunk 12]
```diff
@@ -528,3 +533,96 @@ def _generate_invoice_date(year=datetime.now().year, month=1, this_month=False):
...
+            [],
+            sent_mail.attachments
+        )
+        self.assertEqual(
+            ['to@example.com'],
+            sent_mail.to
+        )
+        self.assertEqual(len(mail.outbox), 1)
+
+    def test_send_trial_instances_report(self):
```
Comment (lgp171188) [inline reports/tests.py:587]: @spokerman12, this test doesn't test the corner cases like the one mentioned [here](https://github.com/open-craft/opencraft/pull/711/files#r580161566).

File: reports/tasks.py (hunk 11 above)
Comment (spokerman12) [reply reports/tasks.py]: Thank you for picking up this case. I've modified the filter and added the unit test

File: reports/tests.py (hunk 12 above)
Comment (spokerman12) [reply reports/tests.py:587]: I added the test case as stated in previous comment

File: reports/tests.py [hunk 13]
```diff
@@ -528,3 +533,160 @@ def _generate_invoice_date(year=datetime.now().year, month=1, this_month=False):
...
+            send_trial_instances_report(recipients=['to@example.com'])
+            sent_mail = mail.outbox[0]
+
+            self.assertEqual(len(mail.outbox), 1)
+
+            self.assertIn(
+                'Please find attached a CSV with the statistics for February 2020',
+                sent_mail.body
+            )
+            print(sent_mail.attachments[0][1])
```
Comment (lgp171188) [inline reports/tests.py:664]: @spokerman12, nit: this should be removed.

Comment (spokerman12) [reply reports/tests.py:664]: My bad!

File: reports/tests.py [hunk 14]
```diff
@@ -528,3 +533,159 @@ def _generate_invoice_date(year=datetime.now().year, month=1, this_month=False):
...
+        """
+        Returns a trial instance for a user
+
+        :param key: String to append to properties.
+        :return: A trial instance
+        """
+        user, _ = get_user_model().objects.get_or_create(
+            username=key + 'test', email=key + 'test@example.com')
+        instance = OpenEdXInstanceFactory(
+            successfully_provisioned=True, betatestapplication=BetaTestApplication())
```
Comment (lgp171188) [inline reports/tests.py:553]: @spokerman12, nit: formatting of closing parenthesis doesn't match the coding standards that we follow.

Comment (spokerman12) [reply reports/tests.py]: Fixed plus another occurrence.

File: reports/tests.py [hunk 15]
```diff
@@ -528,3 +533,161 @@ def _generate_invoice_date(year=datetime.now().year, month=1, this_month=False):
...
+        """
+        Tests that if send_trial_instances_report succeeds, the recipient receives
+        a correct report with only instances created last month.
+        """
+        date1 = "2020-01-01"
+        date2 = "2020-02-01"
+        date3 = "2020-03-01"
+
+        with freeze_time(date1):
+            instance1 = self.create_user_with_trial_instance(key='1')
```
Comment (lgp171188) [inline reports/tests.py:599]: @spokerman12, the key doesn't matter as it is not used for any of the checks below. So can't the `create_user_with_trial_instance()` method be a factory method that accepts a count and returns the `count` number of instances, which can be assigned like

```python
instance1, instance2, instance3 = self.user_trial_instance_factory(count=3)
```
?

Comment (spokerman12) [reply reports/tests.py]: True. Modified as requested.

**Review by lgp171188 — APPROVED** [review]
@spokerman12, :+1: Please wait till the unrelated CI failures are fixed before merging this PR.
- [x] I tested and verified that the periodic job now includes only the instances created in the last completed month.
- [x] I read through the code
- [x] ~~I checked for accessibility issues~~ NA
- [x] ~~Includes documentation~~ NA

**Comment by lgp171188** [comment]
@spokerman12, can you fix the failing quality check before moving the task to blocker?

**Comment by spokerman12** [comment]
@lgp171188 sure. I'd be in favor of adding a lint-fix for python; I don't believe it's implemented.
